x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/changepoint/ChangePointAggregator.java

```diff
@@ -143,502 +71,35 @@ public InternalAggregation doReduce(InternalAggregations aggregations, Aggregati
-        // before we run the tests.
-        SampleData sampleData = sample(values, weights, discoveredChangePoints);
-        final double[] sampleValues = sampleData.values();
-        final double[] sampleWeights = sampleData.weights();
-
-        double pValue = 1;
-        for (int cp : sampleData.changePoints()) {
-            double[] x = Arrays.copyOfRange(sampleValues, 0, cp);
-            double[] y = Arrays.copyOfRange(sampleValues, cp, sampleValues.length);
-            double statistic = KOLMOGOROV_SMIRNOV_TEST.kolmogorovSmirnovStatistic(x, y);
-            double ksTestPValue = KOLMOGOROV_SMIRNOV_TEST.exactP(statistic, x.length, y.length, false);
-            if (ksTestPValue < pValue) {
-                changePoint = cp;
-                pValue = ksTestPValue;
-            }
-        }
-
-        // We start to get false positives if we have too many candidate change points. This
-        // is the classic p-value hacking problem. However, the Sidak style correction we use
-        // elsewhere is too conservative because test statistics for different split positions
-        // are strongly correlated. We assume that we have some effective number of independent
-        // trials equal to f * n for f < 1. Simulation shows the f = 1/50 yields low Type I
-        // error rates.
-        pValue = independentTrialsPValue(pValue, (sampleValues.length + 49) / 50);
-        logger.trace("distribution change p-value: [{}]", pValue);
-
-        return new TestStats(Type.DISTRIBUTION_CHANGE, pValue, changePoint, stats);
-    }
-
-    enum Type {
-        STATIONARY,
-        NON_STATIONARY,
-        STEP_CHANGE,
-        TREND_CHANGE,
-        DISTRIBUTION_CHANGE
-    }
-
-    record TestStats(Type type, double pValue, double var, double nParams, int changePoint, DataStats dataStats) {
-        TestStats(Type type, double pValue, int changePoint, DataStats dataStats) {
-            this(type, pValue, 0.0, 0.0, changePoint, dataStats);
-        }
-
-        TestStats(Type type, double pValue, double var, double nParams, DataStats dataStats) {
-            this(type, pValue, var, nParams, -1, dataStats);
-        }
-
-        boolean accept(double pValueThreshold) {
-            // Check the change is:
-            // 1. Statistically significant.
-            // 2. That we explain enough of the data variance overall.
-            return pValue < pValueThreshold && rSquared() >= 0.5;
-        }
-
-        double rSquared() {
-            return 1.0 - var / dataStats.var();
-        }
-
-        double pValueVsStationary() {
-            return independentTrialsPValue(
-                fTestNestedPValue(dataStats.nValues(), dataStats.var(), 1.0, var, nParams),
-                dataStats.nCandidateChangePoints()
+    static ChangeType getChangeType(MlAggsHelper.DoubleBucketValues bucketValues) {
+        if (bucketValues.getValues().length < (2 * MINIMUM_BUCKETS) + 2) {
```

Comment:
I wonder if it makes sense to migrate this condition and `MINIMUM_BUCKETS` to `ChangeDetector`, since it's tied up with the tests that class performs and we'll probably want the same condition for ES|QL.

Reply:
This `static ChangeType ChangePointAggregator::getChangeType` is currently the main entrypoint, that ES|QL will also use (after making it public). Maybe it's better to move it to a separate class `ChangePointDetector` or so. WDYT?

Reply:
FYI, moved it as I suggested